File: frontend/src/lib/manual.ts
// User Manual data access (Stage 7a, doc 21 §7/§8): the all-role Published
// reader stream + server-side search, and the Admin-only publish/replace/
// delete/restore mutations. Every write is Admin-gated at the ROUTE and again
// in the SERVICE (require_manual_admin / require_trash_admin) — UI
// hide/disable is never authorization (doc 21 §2); a non-Admin sees the
// server 403 envelope verbatim.
//
// EMPIRICAL route map (signatures read — PR #105/#111/#113 lesson):
// - GET /manual/stream + GET /manual/search: all-role reads, keyset cursor,
//   meta carries the stream_version the page/anchors resolve against.
// - POST /admin/manual/documents (+ :upload): OCC is the BODY-form
//   `expected_stream_version` INT (optional server-side; this client always
//   guards with the rendered stream snapshot — UM-13/UM-15) + a fresh
//   Idempotency-Key per attempt.
// - POST .../{id}/revisions: OCC is the BODY-form `expected_head_revision_id`
//   STRING (the route prefers the body over the If-Match transport).
// - DELETE .../{id}: carries an OPTIONAL BODY (reason + expected_stream_version)
//   + Idempotency-Key — api.del takes neither, so apiRequest is used directly.
// - POST .../{id}:restore: NO body, Idempotency-Key only; gated by
//   require_trash_admin (NOT manual admin) and delegates to the landed Trash
//   restore — the return shape IS lib/trash.ts RestoreResult.
//
// The manual has no dedicated SSE event: reads live under ["manual"] and are
// swept by resource.changed. GET /manual/section is NOT routed (doc 21 §12 —
// the Agent reads sections through the Tool Gateway, not HTTP).

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api, apiRequest } from "./apiClient";
import type { RestoreResult } from "./trash";

// ---------------------------------------------------------------------------
// Wire types (mirror application/queries/manual.py and commands/manual.py
// return dicts verbatim)
// ---------------------------------------------------------------------------

// Canonical safe-render block (doc 21 §9.2). payload shapes by block_type:
// heading {level, text} / paragraph {text} / bullet_list|ordered_list {items}
// / code {code_text, language} / callout {tone, title, text} / divider {}.
export interface ManualBlock {
  block_id: string;
  block_type: string;
  anchor: string;
  payload: Record<string, unknown>;
}

export interface ManualSection {
  document_id: string;
  is_baseline: boolean;
  title: string;
  revision_id: string;
  revision_no: number;
  source_type: string;
  source_label: string;
  stream_position: number;
  anchor: string;
  blocks: ManualBlock[];
}

export interface ManualStreamPage {
  data: ManualSection[];
  meta: {
    stream_version: number;
    cursor: string | null;
    has_more: boolean;
    limit: number;
  };
}

export interface ManualSearchResult {
  chunk_id: string;
  document_id: string;
  revision_id: string;
  revision_no: number;
  title: string;
  heading_path: string;
  excerpt: string;
  anchor: string;
  block_ids: string[];
  source_label: string;
}

export interface ManualSearchPage {
  data: ManualSearchResult[];
  meta: {
    stream_version: number;
    cursor: string | null;
    has_more: boolean;
    limit: number;
    query: string;
  };
}

// commands/manual.py _publish_new_document return (create + upload share it).
export interface PublishResult {
  document_id: string;
  revision_id: string;
  revision_no: number;
  stream_entry_id: string;
  stream_position: number;
  anchor: string;
  title: string;
  checksum: string;
  stream_version: number;
  correlation_id: string | null;
}

export interface ReviseResult {
  document_id: string;
  revision_id: string;
  revision_no: number;
  superseded_revision_id: string;
  stream_position: number;
  anchor: string;
  stream_version: number;
  correlation_id: string | null;
}

// Two server branches share this shape: the idempotent-repeat branch adds
// idempotent_repeat and omits display_name.
export interface DeleteResult {
  document_id: string;
  deletion_state: string;
  trash_entry_id: string | number | null;
  stream_version: number;
  correlation_id: string | null;
  display_name?: string;
  idempotent_repeat?: boolean;
}

// ---------------------------------------------------------------------------
// Taxonomy mirrors (hydration/presentation only — the server re-validates
// every dispatch; a rejected extension is MANUAL_FILE_TYPE_UNSUPPORTED
// verbatim, doc 21 §5 UM-06)
// ---------------------------------------------------------------------------

// domain/manual/enums.py SOURCE_TYPE_BY_EXTENSION keys, same order.
export const ACCEPTED_UPLOAD_EXTENSIONS = [".txt", ".md", ".markdown", ".html", ".htm"] as const;

// ---------------------------------------------------------------------------
// Query hooks — all under the ["manual"] prefix (swept by resource.changed)
// ---------------------------------------------------------------------------

export function useManualStream(cursor: string | null) {
  return useQuery({
    queryKey: ["manual", "stream", cursor],
    queryFn: () => {
      const params = new URLSearchParams();
      if (cursor !== null) params.set("cursor", cursor);
      const qs = params.toString();
      return api.get<ManualStreamPage>(`/manual/stream${qs ? `?${qs}` : ""}`);
    },
    // Keep the current stream mounted while a page flip is in flight.
    placeholderData: (previous) => previous,
  });
}

// A blank query searches nothing server-side (doc 21 §14) — the hook stays
// disabled until there is a needle, so no useless round-trip is made.
export function useManualSearch(q: string, cursor: string | null) {
  const needle = q.trim();
  return useQuery({
    queryKey: ["manual", "search", needle, cursor],
    queryFn: () => {
      const params = new URLSearchParams();
      params.set("q", needle);
      if (cursor !== null) params.set("cursor", cursor);
      return api.get<ManualSearchPage>(`/manual/search?${params.toString()}`);
    },
    enabled: needle.length > 0,
    placeholderData: (previous) => previous,
  });
}

// ---------------------------------------------------------------------------
// Mutations — Admin-only server-side; fresh Idempotency-Key per attempt so a
// retry after a rejection is a new decision, not a replay
// ---------------------------------------------------------------------------

export interface CreateDocumentInput {
  title: string;
  content: string;
  allow_duplicate: boolean;
  expected_stream_version: number;
}

export function useCreateManualDocument() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: CreateDocumentInput) =>
      apiRequest<PublishResult>("/admin/manual/documents", {
        method: "POST",
        headers: { "Idempotency-Key": crypto.randomUUID() },
        body: input,
      }),
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: ["manual"] });
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
    },
  });
}

export interface UploadDocumentInput {
  source_filename: string;
  content: string;
  // Omitted when blank: the server derives the title from the filename stem.
  title?: string;
  allow_duplicate: boolean;
  expected_stream_version: number;
}

export function useUploadManualDocument() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: UploadDocumentInput) =>
      apiRequest<PublishResult>("/admin/manual/documents:upload", {
        method: "POST",
        headers: { "Idempotency-Key": crypto.randomUUID() },
        body: input,
      }),
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: ["manual"] });
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
    },
  });
}

// NOTE (doc 21 §7): revision replacement is a landed backend capability the
// V18 reader spec does not surface; this client binds it as an explicit Admin
// maintenance affordance so the full route surface is frontend-bound (the
// PR #95 gated-POST precedent). The server still gates it end-to-end.
export interface ReplaceRevisionInput {
  document_id: string;
  content: string;
  // Omitted when blank: the server keeps the current head title.
  title?: string;
  // The section's visible head revision id — a moved head is a 409
  // MANUAL_REVISION_CONFLICT verbatim (doc 21 §7/§10; body wins over If-Match).
  expected_head_revision_id: string;
}

export function useReplaceManualRevision() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ document_id, ...body }: ReplaceRevisionInput) =>
      apiRequest<ReviseResult>(
        `/admin/manual/documents/${encodeURIComponent(document_id)}/revisions`,
        {
          method: "POST",
          headers: { "Idempotency-Key": crypto.randomUUID() },
          body,
        },
      ),
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: ["manual"] });
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
    },
  });
}

export interface DeleteDocumentInput {
  document_id: string;
  // Omitted when blank — the wire field is nullable.
  reason?: string;
  // Stale snapshot -> 409 MANUAL_STREAM_CONFLICT verbatim (UM-15).
  expected_stream_version: number;
}

// Soft delete writes a page-contract Trash entry in the same tx (doc 21 §8.4)
// — the Admin Trash index moves, so ["trash"] is invalidated alongside.
export function useSoftDeleteManualDocument() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ document_id, ...body }: DeleteDocumentInput) =>
      apiRequest<DeleteResult>(`/admin/manual/documents/${encodeURIComponent(document_id)}`, {
        method: "DELETE",
        headers: { "Idempotency-Key": crypto.randomUUID() },
        body,
      }),
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: ["manual"] });
      void queryClient.invalidateQueries({ queryKey: ["trash"] });
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
    },
  });
}

// Restore is addressed by DOCUMENT id and delegates to the landed Trash core
// (doc 21 §8.4 UM-09): no body, no OCC token — the recoverable Trash entry is
// resolved server-side; the return shape is the Trash RestoreResult.
export function useRestoreManualDocument() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (input: { document_id: string }) =>
      apiRequest<RestoreResult>(
        `/admin/manual/documents/${encodeURIComponent(input.document_id)}:restore`,
        {
          method: "POST",
          headers: { "Idempotency-Key": crypto.randomUUID() },
        },
      ),
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: ["manual"] });
      void queryClient.invalidateQueries({ queryKey: ["trash"] });
      void queryClient.invalidateQueries({ queryKey: ["audit"] });
    },
  });
}

File: backend/src/entropia/apps/api/routes/manual.py
"""User Manual endpoints (Stage 7a, doc 21 §7).

Thin handlers: parse body/headers -> one application command/query. Published
stream + search are all-role (Admin/Supervisor/User; the Agent uses the Tool
Gateway, not these routes). Every write route re-checks Admin at the ROUTE and
the command re-checks it at the SERVICE (``require_manual_admin`` — UI
hide/disable is never authorization, doc 21 §2). Concurrency: stream mutations
take ``expected_stream_version``; revision replacement takes
``expected_head_revision_id`` (transport: ``If-Match``); every mutation carries
an ``Idempotency-Key`` (doc 21 §7).
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, File, Form, Header, Query, UploadFile
from pydantic import BaseModel

from entropia.application.commands.manual import (
    create_manual_document,
    replace_manual_revision,
    restore_manual_document,
    soft_delete_manual_document,
    upload_manual_document,
)
from entropia.application.queries.manual import get_manual_stream, search_manual
from entropia.apps.api.deps import RequestContext, request_context
from entropia.apps.api.upload import validate_multipart_upload
from entropia.domain.identity.policy import require_manual_admin, require_trash_admin

router = APIRouter(tags=["manual"])

_STREAM_PATH = "/manual/stream"
_SEARCH_PATH = "/manual/search"
_DOCUMENTS_PATH = "/admin/manual/documents"
_UPLOAD_PATH = "/admin/manual/documents:upload"
_DOCUMENT_PATH = "/admin/manual/documents/{document_id}"
_REVISIONS_PATH = "/admin/manual/documents/{document_id}/revisions"
_RESTORE_PATH = "/admin/manual/documents/{document_id}:restore"


class CreateDocumentRequest(BaseModel):
    title: str
    content: str
    allow_duplicate: bool = False
    expected_stream_version: int | None = None


class ReplaceRevisionRequest(BaseModel):
    content: str
    title: str | None = None
    expected_head_revision_id: str | None = None


class DeleteDocumentRequest(BaseModel):
    reason: str | None = None
    expected_stream_version: int | None = None


def _expected_revision(body_value: str | None, if_match: str | None) -> str | None:
    """Body token wins; ``If-Match`` transports the same head revision id."""
    if body_value is not None:
        return body_value
    if if_match is None:
        return None
    return if_match.strip().strip('"') or None


@router.get(_STREAM_PATH)
async def stream(
    cursor: str | None = Query(default=None),
    limit: int | None = Query(default=None, ge=1, le=100),
    ctx: RequestContext = Depends(request_context),
) -> dict[str, Any]:
    return await get_manual_stream(ctx.session, ctx.actor, cursor=cursor, limit=limit)


@router.get(_SEARCH_PATH)
async def search(
    q: str | None = Query(default=None),
    cursor: str | None = Query(default=None),
    limit: int | None = Query(default=None, ge=1, le=100),
    ctx: RequestContext = Depends(request_context),
) -> dict[str, Any]:
    return await search_manual(ctx.session, ctx.actor, q=q, cursor=cursor, limit=limit)


@router.post(_DOCUMENTS_PATH, status_code=201)
async def create_document(
    body: CreateDocumentRequest,
    ctx: RequestContext = Depends(request_context),
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
) -> dict[str, Any]:
    require_manual_admin(ctx.actor)
    return await create_manual_document(
        ctx.session,
        ctx.actor,
        title=body.title,
        content=body.content,
        allow_duplicate=body.allow_duplicate,
        expected_stream_version=body.expected_stream_version,
        idempotency_key=idempotency_key,
    )


@router.post(_UPLOAD_PATH, status_code=201)
async def upload_document(
    file: UploadFile = File(...),
    title: str | None = Form(default=None),
    allow_duplicate: bool = Form(default=False),
    expected_stream_version: int | None = Form(default=None),
    ctx: RequestContext = Depends(request_context),
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
) -> dict[str, Any]:
    """Real native document upload (F-03): the browser transfers the selected
    UTF-8 TXT/MD/HTML file as ``multipart/form-data``. Size and UTF-8 encoding
    are validated server-side; the TXT/MD/HTML extension gate lives in the
    command (``MANUAL_FILE_TYPE_UNSUPPORTED``). The reader-visible title is
    filename-derived unless an explicit override is supplied."""
    require_manual_admin(ctx.actor)
    upload = await validate_multipart_upload(file)
    return await upload_manual_document(
        ctx.session,
        ctx.actor,
        source_filename=upload.filename or "",
        content=upload.text,
        title=title,
        allow_duplicate=allow_duplicate,
        expected_stream_version=expected_stream_version,
        idempotency_key=idempotency_key,
    )


@router.post(_REVISIONS_PATH, status_code=201)
async def replace_revision(
    document_id: str,
    body: ReplaceRevisionRequest,
    ctx: RequestContext = Depends(request_context),
    if_match: str | None = Header(default=None, alias="If-Match"),
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
) -> dict[str, Any]:
    require_manual_admin(ctx.actor)
    return await replace_manual_revision(
        ctx.session,
        ctx.actor,
        document_id=document_id,
        content=body.content,
        title=body.title,
        expected_head_revision_id=_expected_revision(body.expected_head_revision_id, if_match),
        idempotency_key=idempotency_key,
    )


@router.delete(_DOCUMENT_PATH)
async def soft_delete_document(
    document_id: str,
    body: DeleteDocumentRequest | None = None,
    ctx: RequestContext = Depends(request_context),
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
) -> dict[str, Any]:
    require_manual_admin(ctx.actor)
    payload = body or DeleteDocumentRequest()
    return await soft_delete_manual_document(
        ctx.session,
        ctx.actor,
        document_id=document_id,
        reason=payload.reason,
        expected_stream_version=payload.expected_stream_version,
        idempotency_key=idempotency_key,
    )


@router.post(_RESTORE_PATH)
async def restore_document(
    document_id: str,
    ctx: RequestContext = Depends(request_context),
    idempotency_key: str | None = Header(default=None, alias="Idempotency-Key"),
) -> dict[str, Any]:
    require_trash_admin(ctx.actor)
    return await restore_manual_document(
        ctx.session, ctx.actor, document_id=document_id, idempotency_key=idempotency_key
    )


__all__ = ["router"]
